sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningDeserializer.java

```diff
@@ -65,58 +68,185 @@ public static SimpleModule getModule(final ObjectMapper mapper) {
+            } else {
+                final JsonTypeInfo typeInfo = c.getAnnotation(com.fasterxml.jackson.annotation.JsonTypeInfo.class);
+                if (typeInfo != null) {
+                    String typeId = typeInfo.property();
+                    if (containsDot(typeId)) {
+                        final String typeIdOnWire = unescapeEscapedDots(typeId);
+                        JsonNode typeIdValue = ((ObjectNode) currentJsonNode).remove(typeIdOnWire);
+                        if (typeIdValue != null) {
+                            ((ObjectNode) currentJsonNode).put(typeId, typeIdValue);
                         }
-                        ((ObjectNode) root).set(value, node);
                     }
                 }
             }
         }
-        JsonParser parser = new JsonFactory().createParser(root.toString());
-        parser.nextToken();
-        return defaultDeserializer.deserialize(parser, ctxt);
+        return tDeserializer.deserializeTypedFromAny(newJsonParserForNode(currentJsonNode), cxt);
+    }
+
+    @Override
+    public Object deserialize(JsonParser jp, DeserializationContext cxt) throws IOException {
+        // This method will be called by Jackson for each "Json object" in the input wire stream
+        // it is trying to deserialize.
+        //
+        // The parameter 'jp' is the reader to read "Json object with TypeId"
+        //
+        JsonNode currentJsonNode = mapper.readTree(jp);
+        if (currentJsonNode.isNull()) {
+            currentJsonNode = mapper.getNodeFactory().objectNode();
+        }
+        final Class<?> tClass = this.defaultDeserializer.handledType();
+        for (Class<?> c : TypeUtil.getAllClasses(tClass)) {
+            if (c.isAssignableFrom(Object.class)) {
+                continue;
+            } else {
+                for (Field classField : c.getDeclaredFields()) {
+                    handleFlatteningForField(classField, currentJsonNode);
+                }
+            }
+        }
+        return this.defaultDeserializer.deserialize(newJsonParserForNode(currentJsonNode), cxt);
     }
 
     @Override
-    public void resolve(DeserializationContext ctxt) throws JsonMappingException {
-        ((ResolvableDeserializer) defaultDeserializer).resolve(ctxt);
+    public void resolve(DeserializationContext cxt) throws JsonMappingException {
+        ((ResolvableDeserializer) this.defaultDeserializer).resolve(cxt);
+    }
+
+    /**
+     * Given a field of a POJO class and JsonNode corresponds to the same POJO class,
+     * check field's {@link JsonProperty} has flattening dots in it if so
+     * flatten the nested child JsonNode corresponds to the field in the given JsonNode.
+     *
+     * @param classField the field in a POJO class
+     * @param jsonNode the json node corresponds to POJO class that field belongs to
+     */
+    @SuppressWarnings("unchecked")
+    private static void handleFlatteningForField(Field classField, JsonNode jsonNode) {
```

Comment:
Should we just use `ObjectNode`?

Reply:
I'm not sure we should do the upcasting ahead - w.r.t `jsonProperty` lookup and modification in this method we're trying not to move away from the current logic (in released versions) to mutate ObjectNode. The `ObjectNode` is type extending from `JsonNode` and today we do the upcasting only if we find it has `jsonProperty`, but if we have to take `ObjectNode` here then we had to do the upcasting for all nodes from the caller. 

An example is - In calling side, when we attempted to assign the value of readTree to `ObjectNode`, there happened a runtime cast exception where readTree returned `NullNode` which is not an `ObjectNode` but `JsonNode`, and auto-casting based on receiving variable type failed.

Reply:
For reference: upcasting in released versions [here](https://github.com/Azure/azure-sdk-for-java/blob/4a22178c595320eb1fb658d13dae0591154a9b73/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningDeserializer.java#L106), see line-91  where we do check for annotation presence before the upcast, the same pattern followed in this PR.